---
slug: /basics/failures
sidebar_position: 6
title: Failures
---

## What is a failure?

Errors prevent the achievement of the service goal due to unsatisfied validations or exceptions.

At the same time, _failures_ prevent the achievement of the service goal due to business reasons.

That is why is so vital to use meaningful names for services.

For instance, when the aim is <u>to order a bus ticket</u>, call it - `OrderBusTicket`.

So when the service is invoked, but the ticket is not ordered for some business reason, it should be treated as a failure.

Why? Because business reason is not an error, since the input is valid, the environment is stable, but the service goal is still not achieved.

Please, read the following code snippet carefully to get the idea.

```ruby
class OrderBusTicket
  # ...
  def result
    return error("Bus `#{bus.number}` is NOT valid") unless bus.valid?

    return failure("Bus has no available seats") if bus.full?
    return failure("Bus trip is postponed") if ::Trip.for_bus(bus).postponed?

    ticket = ::BusTerminal.issue_ticket(bus)

    success(ticket: ticket)
  end
end
```

And this is how it can be triggerred:

```ruby
result = OrderBusTicket.result(bus: bus)

if result.failure?
  # `result.message` is a string-like object with failure reason.
  result.message
end
```

When the bus has not valid number we can not even try to order a ticket.

Hence the error is returned as the result.

But once all the error preconditions are met, the ticket can still be not ordered due to the business reasons like lack of available seats or the trip delay.

That is why those cases are examples of failures.
